### ISSUE TITLE
Geofencing API - Adding capability to manage max number of notification

### ISSUE THREAD

**Issue by bigludo7** [camaraproject/DeviceLocation#111]
**Problem description**
As of now we are only able to subscribe for a period of time but why not allowing also to subscribe to a number of notification?

We can have UC where we want to get notified when a device will reach or exit a geo zone once but without any time limit (like for sim card that are embedded in equipment that are not suppose to move like in traffic light...).

**Possible evolution**
Add capability to define a max number of notifications.

**Alternative solution**

**Additional context**

**Comment by jlurien** [comment]
I don't fully understand the UC and the suggestion. Is that to set a limit to the number of notifications until `subscriptionExpireTime`, for certain device? 

In case of a device with a fixed location, like a traffic light, I guess that the interesting use case is to detect if the device is being abnormally moved, e.g. stolen, right? In that case you would receive a single notification when leaving the geofence

**Comment by bigludo7** [comment]
@jlurien

> Is that to set a limit to the number of notifications until subscriptionExpireTime, for certain device?

Yes. If the `subscriptionExpireTime`  is provided, the subscriptions will stop when this date is reached or the max number of notification is reached. If the  `subscriptionExpireTime`  is not provided then the subscriptions will stop when the max number of notification is reached. There is same feature in 3GPP monitoringEvent with the `maximumNumberOfReports` attribute.


> In case of a device with a fixed location, like a traffic light, I guess that the interesting use case is to detect if the device is being abnormally moved, e.g. stolen, right? In that case you would receive a single notification when leaving the geofence

Yes this is one of the UC. I was thinking that the expirationTime is not really the relevant request parameter as the expectation is to get notified when the device entered (or exited) the defined area whenever it is it is in one week, one month or 10 years...

One way to solve this perhaps is to consider that if `subscriptionExpireTime` is not provided then only one notification will be sent.

**Comment by jlurien** [comment]
This sort of limitation sounds like something generic, not specific to Geofencing, maybe it's convenient to to discuss it at Commonalities level, as `subscriptionExpireTime` which comes from the guidelines. Only that it is not part of the CloudEvents format, so we would need something within the payload.

My assumption regarding `subscriptionExpireTime`, which is optional, was that if not provided there is no explicit time limit (other than that supported by implementations, that may be returned in the response). I think it is not a good idea to assume that only one event is to be sent in this case. If needed, better to have some dedicated attribute.

**Comment by alpaycetin74** [comment]
My understanding so far is, if a device leaves the subscribed area, the API sends a notification to inform the user it has happened. There is only one notification. Even if the device _continues to remain outside the area_, no further notifications are sent.
With that approach, a notification can be sent _again_ only if the device _enters_ the area again and _leaves_ again.

**Comment by jlurien** [comment]
> My understanding so far is, if a device leaves the subscribed area, the API sends a notification to inform the user it has happened. There is only one notification. Even if the device _continues to remain outside the area_, no further notifications are sent. With that approach, a notification can be sent _again_ only if the device _enters_ the area again and _leaves_ again.

My understanding on this is the same

**Comment by bigludo7** [comment]
@alpaycetin74 @jlurien  Agreed....  but it did not solve my point.

If I need to be notified once that a device entered or existed a zone what is the point to get the subscription valid because the subcriptionExpiryTime is not reached ? My request is to be able to get notify once and get the subscription terminated once the event occured.

**Comment by jlurien** [comment]
I suggest to move the requirement to Commonalities to discuss the use case and, eventually, how to model a limit to the amount of events to be generated per subscription. We have also to define the expected logic when the limit is reached, do we expect the subscription to terminate as with expired subscriptions? In that case a subscription-ends event should be triggered with a new `terminationReason: MAX_EVENTS_REACHED` or similar.

**Comment by bigludo7** [comment]
Thanks - Fully inline with your proposal @alpaycetin74 and for the suggested way to proceed @jlurien. Will do the Commonalities issue and copy/paste @alpaycetin74 proposal.

**Comment by jlurien** [comment]
Following discussion in Commonalities.

**Comment by bigludo7** [comment]
Now that the `subscriptionMaxEvents` has been added in the guideline we can add it in our Geofencing API (https://github.com/camaraproject/Commonalities/pull/131) 

@maxl2287 as you are the active contributor on this API, is it something you can consider ?
Thanks !
